glean-core/src/metrics/labeled.rs

```diff
@@ -184,32 +100,98 @@ where
-        // We have 3 scenarios to consider:
+    pub fn get(&mut self, label: &str) -> T {
+        // We have 2 scenarios to consider:
         // * Static labels. No database access needed. We just look at what is in memory.
-        // * Dynamic labels, initialized Glean. We look up in the database all previously stored
-        //   labels in order to keep a maximum of allowed labels.
-        // * Dynamic labels, uninitialized Glean. We ignore potentially previously stored labels
-        //   and just take what the user passed in.
-        //   This potentially allows creating more than `MAX_LABELS` labels, but only before Glean
-        //   is initialized.
-        //   This behavior is not publicly documented and should not be abused/depended upon.
-        // The last case is buggy behavior, tracked in bug 1588451.
-        let glean = glean.into();
-        let label = match (&self.labels, glean) {
-            (Some(_), _) => self.static_label(label),
-            (None, Some(glean)) => self.dynamic_label(glean, label),
-            (None, None) => {
-                if let Ok(label) = validate_label(label) {
-                    label
-                } else {
-                    // We can't report any error, as we don't have a Glean instance.
-                    OTHER_LABEL
-                }
+        // * Dynamic labels. We look up in the database all previously stored
+        //   labels in order to keep a maximum of allowed labels. This is done later
+        //   when the specific metric is actually recorded, when we are guaranteed to have
+        //   an initialized Glean object.
+        match self.labels {
+            Some(_) => {
+                let label = self.static_label(label).to_string();
+                self.new_metric_with_name(combine_base_identifier_and_label(
+                    &self.submetric.meta().name,
+                    &label,
+                ))
             }
-        };
-        let label = format!("{}/{}", self.submetric.meta().name, label);
+            None => self.new_metric_with_dynamic_label(label.to_string()),
+        }
+    }
+}
+
+/// Combines a metric's base identifier and label
+pub fn combine_base_identifier_and_label(base_identifer: &str, label: &str) -> String {
+    format!("{}/{}", base_identifer, label)
+}
+
+/// Strips the label off of a complete identifier
+pub fn strip_label(identifier: &str) -> &str {
+    // safe unwrap, first field of a split always valid
+    identifier.splitn(2, '/').next().unwrap()
+}
+
+/// Validate a dynamic label, changing it to OTHER_LABEL if it's invalid.
+///
+/// Checks the requested label against limitations, such as the label length and allowed
+/// characters.
+///
+/// ## Arguments
+///
+/// * `label` - The requested label
+///
+/// ## Return value
+///
+/// Returns the entire identifier for the metric, including the base identifier and the
+/// corrected label.
+/// The errors are logged.
+pub fn dynamic_label(
+    glean: &Glean,
```

Comment:
Looking here and in the database implementation again: Do we really need `Glean`? Or do we just need `&Database`? This way the database can pass `self` and everyone else passed `glean.storage()` and `.record` doesn't need to pass a new thing at all

Reply:
That's a killer suggestion! I like it! And would definitely make it easier to digest touching so many files :)

Reply:
We need `glean` to record errors.  (Unless we refactored the error recording code to only require a database, but that would mean we couldn't use "real" labeled counters for error recording, we'd have to sort of duplicate that code directly on top of the database API).

Reply:
*goes off to rework the whole error reporting system* Hmmmmmm

Reply:
So consensus is to refactor the error recording code to make this possible?

Reply:
uhm, I'd rather not. Not before merging to a-c. That would be a significant change (and I don't even know if it's possible).

One other thing would be to return a `Result<String, (String, String)>` all the way up and do the error recording in another place, but that also seems ugly and will need to touch a couple of places to make it work.

Reply:
I think for now we don't have a better way than to keep it like you did, file a bug and take a look at it again in a week or two once we're more open for bigger changes again.

Reply:
(This is all under the premise that we want to get this in before a first release next week, as one of the last big changes, this kinda affects the current QA process)

Reply:
Ok.  Fair enough.  Filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1589675)